# IMS_3.04
- **Description** : Etude de deux variables qualitatives
- **Source** : Introduction à la méthode statistique - Catherine Pardoux  
- **Chapitre** :  3 - Distributions statistiques à deux caractères
- **Exercice** : 3.4

## Enoncé
![Enonce IMS_3.04](../images/Enonces/IMS_3.04_enonce.PNG)

**Question supplémentaire**   
Evaluer la dépendance des deux variables avec les méthodes suivantes :
- Méthode 1 : comparaison des effectifs théoriques et effectifs observés
- Méthode 2 : Méthode du khi-deux
- Méthode 3 : Méthode du phi-deux
- Méthode 4 : Coefficient de Cramer   

Pour cela commencer par réaliser le tableau des effectifs théoriques (cad tableau des valeurs conjointes de X et Y selon l'hypothèse d'indépendance)

## Chargement des données

In [59]:
import pandas as pd
import numpy as np
import math

DATA = "IMS_3.04_donnees.csv"

In [20]:
donnees = pd.read_csv(DATA, index_col=0)
donnees

Pas du tout satisfait  \
Clients de plus de 2 ans d'ancienneté                     10   
Clients d'au plus 2 ans d'ancienneté                      40   

                                       Plutôt pas satisfait  Plutôt satisfait  \
Clients de plus de 2 ans d'ancienneté                    50               245   
Clients d'au plus 2 ans d'ancienneté                     90               205   

                                       Très satisfait  
Clients de plus de 2 ans d'ancienneté             195  
Clients d'au plus 2 ans d'ancienneté              165

## Question 1

### Calcul de l'effectif total

In [21]:
donnees_totaux = donnees.copy()
#Total sum per column: 
donnees_totaux.loc['Total Y',:] = donnees_totaux.sum(axis=0)
#Total sum per row: 
donnees_totaux.loc[:,'Total X'] = donnees_totaux.sum(axis=1)
donnees_totaux

Pas du tout satisfait  \
Clients de plus de 2 ans d'ancienneté                   10.0   
Clients d'au plus 2 ans d'ancienneté                    40.0   
Total Y                                                 50.0   

                                       Plutôt pas satisfait  Plutôt satisfait  \
Clients de plus de 2 ans d'ancienneté                  50.0             245.0   
Clients d'au plus 2 ans d'ancienneté                   90.0             205.0   
Total Y                                               140.0             450.0   

                                       Très satisfait  Total X  
Clients de plus de 2 ans d'ancienneté           195.0    500.0  
Clients d'au plus 2 ans d'ancienneté            165.0    500.0  
Total Y                                         360.0   1000.0

### Tableau de fréquence

In [22]:
tableau_frequence = (donnees_totaux / donnees_totaux['Total X']['Total Y']) * 100
tableau_frequence

Pas du tout satisfait  \
Clients de plus de 2 ans d'ancienneté                    1.0   
Clients d'au plus 2 ans d'ancienneté                     4.0   
Total Y                                                  5.0   

                                       Plutôt pas satisfait  Plutôt satisfait  \
Clients de plus de 2 ans d'ancienneté                   5.0              24.5   
Clients d'au plus 2 ans d'ancienneté                    9.0              20.5   
Total Y                                                14.0              45.0   

                                       Très satisfait  Total X  
Clients de plus de 2 ans d'ancienneté            19.5     50.0  
Clients d'au plus 2 ans d'ancienneté             16.5     50.0  
Total Y                                          36.0    100.0

### Distribution marginale Y

In [23]:
distribution_marginale_y = pd.DataFrame(tableau_frequence.iloc[:-1,:-1].sum(axis=0), donnees.columns, columns=['Total Y'])
distribution_marginale_y = distribution_marginale_y.transpose()
distribution_marginale_y

Pas du tout satisfait  Plutôt pas satisfait  Plutôt satisfait  \
Total Y                    5.0                  14.0              45.0   

         Très satisfait  
Total Y            36.0

### Pourcentage de clients satisfaits : (Y = "Plutôt satisfait" V "Très satisfait")

In [24]:
pourcentage_clients_satisfaits = float(distribution_marginale_y["Plutôt satisfait"] + distribution_marginale_y["Très satisfait"])
pourcentage_clients_satisfaits

81.0

## Question 2

### Distribution conditionnelle : (X = "> 2 ans ancienneté") | (Y = "Plutôt satisfait" V "Très satisfait")

In [25]:
distribution_conditionnelle_x_sachant_y = (tableau_frequence.iloc[0:1,2:4] / pourcentage_clients_satisfaits) * 100
distribution_conditionnelle_x_sachant_y

Plutôt satisfait  Très satisfait
Clients de plus de 2 ans d'ancienneté         30.246914       24.074074

### Pourcentage de clients de plus de 2 ans d'acienneté parmi les clients satisfaits : (X = "> 2 ans ancienneté") | (Y = "Plutôt satisfait" V "Très satisfait")

In [26]:
pourcentage_clients_plus_2ans_satisfaits = float(distribution_conditionnelle_x_sachant_y["Plutôt satisfait"] + distribution_conditionnelle_x_sachant_y["Très satisfait"])
pourcentage_clients_plus_2ans_satisfaits

54.32098765432099

## Question 3

### Tableau des profils en ligne : Distribution conditionnelle (Y | X)

In [27]:
profils_ligne = (tableau_frequence.iloc[0:2].div(tableau_frequence.iloc[0:2]["Total X"], axis=0)) * 100
profils_ligne

Pas du tout satisfait  \
Clients de plus de 2 ans d'ancienneté                    2.0   
Clients d'au plus 2 ans d'ancienneté                     8.0   

                                       Plutôt pas satisfait  Plutôt satisfait  \
Clients de plus de 2 ans d'ancienneté                  10.0              49.0   
Clients d'au plus 2 ans d'ancienneté                   18.0              41.0   

                                       Très satisfait  Total X  
Clients de plus de 2 ans d'ancienneté            39.0    100.0  
Clients d'au plus 2 ans d'ancienneté             33.0    100.0

## Question 4

### Tableau de contingence (X ^ (Y = "Satisfaits" V Y = "Insatisfaits"))

In [28]:
tableau_contingence = donnees.copy()
tableau_contingence["Insatisfaits"] = tableau_contingence["Pas du tout satisfait"] + tableau_contingence["Plutôt pas satisfait"]
tableau_contingence["Satisfaits"] = tableau_contingence["Plutôt satisfait"] + tableau_contingence["Très satisfait"]
tableau_contingence = tableau_contingence.iloc[0:,4:]
tableau_contingence

Insatisfaits  Satisfaits
Clients de plus de 2 ans d'ancienneté            60         440
Clients d'au plus 2 ans d'ancienneté            130         370

## Question 5

### Calcul des effectifs marginaux

In [29]:
tableau_contingence_marginaux = tableau_contingence.copy()
#Total sum per column: 
tableau_contingence_marginaux.loc['.j',:] = tableau_contingence_marginaux.sum(axis=0)
#Total sum per row: 
tableau_contingence_marginaux.loc[:,'i.'] = tableau_contingence_marginaux.sum(axis=1)
tableau_contingence_marginaux


Insatisfaits  Satisfaits      i.
Clients de plus de 2 ans d'ancienneté          60.0       440.0   500.0
Clients d'au plus 2 ans d'ancienneté          130.0       370.0   500.0
.j                                            190.0       810.0  1000.0

### Valeur de contingence si les deux variables étaient indépendantes : X = "clients de plus de 2 ans d'ancienneté" ^ Y = "satisfaits"

***Rappel : Calculer la valeur conjointe de deux variables qualitatives sous hypothèse d'indépendance***   

![Rappel - Moyenne pondérée variable quantitative continue](../images/Rappels/rappel_valeur_conjointe_2_variables_qualitatives_independantes.PNG)   

**En d'autres termes** : Valeur théorique selon hypothèse indépendance = (valeur marginale de X * valeur marginale de Y) / effectif total



Si les deux caractères étaient indépendants, le nombre de clients de plus de 2 ans d'ancienneté satisfaits serait :   

n<sub>ij</sub> = (n<sub>i.</sub> . n <sub>.j</sub>) / n = (500 . 810) / 1000 =  405   

avec :
- i = 1 : clients de plus de 2 ans d'ancienneté 
- j = 2 : clients satisfaits
- n : nombre total de clients


## Question supplémentaire

### Tableau des effectifs théoriques

On généralise le calcul précédent aux autres cellules du tableau.

In [39]:
tableau_effectifs_theoriques = tableau_contingence_marginaux.copy()
n = tableau_contingence_marginaux.loc['.j', 'i.']
tableau_effectifs_theoriques.iloc[0, 0] = (tableau_contingence_marginaux.loc['.j', 'Insatisfaits'] * tableau_contingence_marginaux.loc['Clients de plus de 2 ans d\'ancienneté','i.']) / n
tableau_effectifs_theoriques.iloc[1, 0] = tableau_effectifs_theoriques.iloc[0, 0] 
tableau_effectifs_theoriques.iloc[0, 1] = (tableau_contingence_marginaux.loc['.j', 'Satisfaits'] * tableau_contingence_marginaux.loc['Clients d\'au plus 2 ans d\'ancienneté','i.']) / n
tableau_effectifs_theoriques.iloc[1, 1] = tableau_effectifs_theoriques.iloc[0, 1]
del tableau_effectifs_theoriques['i.']
tableau_effectifs_theoriques.drop(['.j'], inplace = True)
tableau_effectifs_theoriques

Insatisfaits  Satisfaits
Clients de plus de 2 ans d'ancienneté          95.0       405.0
Clients d'au plus 2 ans d'ancienneté           95.0       405.0

***Rappel : Etude de la laison entre deux variables qualitatives (Méthode 1 - Comparaison effectifs théoriques et effectifs observés)***   

![Rappel - Méthode 1 - Effectifs théoriques / effectifs observés ](../images/Rappels/rappel_relation_2_variables_qualitatives_methode1.PNG)   


In [52]:
tableau_ratio = tableau_contingence / tableau_effectifs_theoriques
tableau_ratio

Insatisfaits  Satisfaits
Clients de plus de 2 ans d'ancienneté      0.631579     1.08642
Clients d'au plus 2 ans d'ancienneté       1.368421     0.91358

### Méthode 2 : Méthode du khi-deux

***Rappel : Etude de la laison entre deux variables qualitatives (Méthode 2 - Khi-deux)***   

![Rappel - Méthode 2 - Khi-deux ](../images/Rappels/rappel_relation_2_variables_qualitatives_methode2.PNG)   


In [67]:
tableau_khi_deux = ((tableau_contingence - tableau_effectifs_theoriques)**2) / tableau_effectifs_theoriques
khi_deux = tableau_khi_deux.to_numpy().sum()
khi_deux

31.838856400259907

### Méthode 3 : Méthode du phi-deux


***Rappel : Etude de la laison entre deux variables qualitatives (Méthode 3 - Phi-deux)***   

![Rappel - Méthode 3 - Phi-deux ](../images/Rappels/rappel_relation_2_variables_qualitatives_methode3.PNG)   

**Remarque** : On l'utilise surtout pour des tables 2x2 car alors il est compris entre 0 et 1

In [68]:
n = donnees_totaux.loc['Total Y','Total X']
phi_deux = khi_deux / n
phi_deux

0.03183885640025991

### Méthode 4 : Coefficient de Cramer

***Rappel : Etude de la laison entre deux variables qualitatives (Méthode 4 - Coefficient de Cramer)***    

**A partir du khi deux**

![Rappel - Méthode 4 - Khi-deux ](../images/Rappels/rappel_relation_2_variables_qualitatives_methode4.PNG)   

**A partir du phi deux**   

\begin{align}
C = \sqrt{\frac{\phi ^{2}}{min(p,q) - 1}}
\end{align}

#### A partir du khi deux

In [69]:
p = 2 # nombre de modalités de X
q = 2 # Nombre de modalités de Y


numerateur = khi_deux
denominateur = n * (min(p, q) - 1) 
coeff_cramer = math.sqrt(numerateur/denominateur)
coeff_cramer

0.17843445967710359

#### A partir du phi deux

In [71]:
numerateur = phi_deux
denominateur = min(p, q) - 1
coeff_cramer = math.sqrt(numerateur / denominateur)
coeff_cramer

0.17843445967710359